### DISCUSSION TITLE
difference between snap & state?

### DISCUSSION THREAD

**Discussion by deadcoder0904** [pmndrs/valtio#504] [Q&A] [answered]
i have been using `valtio` for over a year now (absolutely the simplest solution to state management) but still get confused when using `snap` & `state`

this is how i write my store:

```ts
import { proxy, useSnapshot } from 'valtio'

import { IS_DEVELOPMENT } from '@/utils/index'
import { License } from '@/types/index'

class Store {
	allLicenses: License[] = []
	currentLicense: License = {
		id: '',
		name: '',
		used: 0,
		total: 1,
	}

	setCurrentLicense(license: License) {
		this.currentLicense = license
	}

	setLicenses(licenses: License[]) {
		this.allLicenses = licenses
	}
}

export const state = proxy(new Store())
export const useValtioStore = () => useSnapshot(state)

if (IS_DEVELOPMENT) {
	import('valtio/utils').then((utils) => {
		const { devtools } = utils
		devtools(state, {
			name: 'store',
		})
	})
}
```

in the readme, you state that `snap` is for reading & `state` is for mutating.

so if i do `.something = `, i should use `state`?

if i have to call a method `.setSomething()`, then i can use both `state` or `snap`, it doesn't matter right?

in `render()` method, i should use `snap` for rendering lists bcz it can re-render on change, right?

so in theory, just for `=`, i should use `state`, right? rest can be `snap` even if it's `snap.setSomething()`?

am i getting it right?

**Comment by dai-shi** [comment, marked as answer]
> so if i do `.something = `, i should use `state`?

Yes, actually, `snap` is a frozen object, so `snap.something = ` will throw an error.

> if i have to call a method `.setSomething()`, then i can use both `state` or `snap`, it doesn't matter right?

This is a good question. It depends on how `.setSomething()` is implemented. If it doesn't use `this`, it doesn't matter.

> in `render()` method, i should use `snap` for rendering lists bcz it can re-render on change, right?

Right, in `render()` method body (not in a callback or useEffect).

> so in theory, just for `=`, i should use `state`, right?

You should also use `state` for reading outside `render()` body.

> am i getting it right?

Mostly. Yeah, but I see your confusion, and this is the hardest part in educating valitio usage.

**Reply by deadcoder0904** [reply]
> This is a good question. It depends on how .setSomething() is implemented. If it doesn't use this, it doesn't matter.

see my code above. it does use `this` in a `class` so it's a no, right?

> Right, in render() method body (not in a callback or useEffect).

so what's to be used in `useEffect` & `useCallback`? should i use `state` for `=` & `snap` for reading or both will do?

> Mostly. Yeah, but I see your confusion, and this is the hardest part in educating valitio usage.

yeah, valtio is really easy once you understand the difference when to use `snap` & `state`. i believe 1 example covering all edge-cases in the readme would help with having less issues & discussions bcz i think most people are confused for that reason or just point to this issue haha :)

**Reply by dai-shi** [reply]
> see my code above. it does use `this` in a `class` so it's a no, right?

oops, sorry, I overlooked it.

If you do `snap.doSomething()`, `this` inside `doSomething` will be `snap`.
This is not specific to valtio, it's just how JS works.

> so what's to be used in `useEffect` & `useCallback`? should i use `state` for `=` & `snap` for reading or both will do?

For deps, we should use primitive values (string, number, ...), not snap objects.
So,
```js
const Component = () => {
  const snap = useSnapshot(...)
  const { num, str } = snap;
  useEffect(() => {
    ...
  }, [num, str])
  ...
```
would just be fine.

**Comment by santiagopuentep** [comment]
From what I could learn, the "snapshot for reading and state for writing" is wrong.

It's snapshot only to access what you want to be reactive, and always only inside the render function, never inside a callback, and state for everything else (for writing and reading inside callbacks, passing as props, etc.). Snapshot seems to be the dangerous one, so I only use it to trigger reactivity and that's it.

What's not clear to me still is if I can pass a snapshot as a prop. I know that for useMemo it can be a problem, but in some other cases not.

I think the docs need clear examples of dos and don'ts for each case and to remove the "snapshot for reading and state for writing".

**Reply by dai-shi** [reply]
To be a little clear, what's difficult to understand is the snapshot object returned by `useSnapshot()` (which is different from what `snapshot()` returns, which is a plain object that never causes problems) because of render optimization.

> What's not clear to me still is if I can pass a snapshot as a prop.

Yes, but if your child component uses `React.memo`, it may behave unexpectedly like `useMemo` and `useEffect` deps.

> remove the "snapshot for reading and state for writing".

Do we still have it?
We develop [eslint-plugin-valtio](https://github.com/pmndrs/eslint-plugin-valtio) for dos and don'ts.

**Reply by santiagopuentep** [reply]
> Do we still have it?

Yes. Here: https://valtio.pmnd.rs/docs/basic/useSnapshot
It says "Rule of thumb: read from snapshots, mutate the source.".

In order to learn Valtio I had to read all issues and discussions to find the ins and outs. There are many important usage tips that are not in the docs so I found it impossible to use in some cases without reading all the issues. With a more example based documentation I think it would be much, much easier for newer devs.

**Reply by deadcoder0904** [reply]
> remove the "snapshot for reading and state for writing".

weird, i found this bit more helpful than others as i followed it to the "t" & wrote the doc.

> In order to learn Valtio I had to read all issues and discussions to find the ins and outs. There are many important usage tips that are not in the docs so I found it impossible to use in some cases without reading all the issues. With a more example based documentation I think it would be much, much easier for newer devs.

💯 i even wrote the same thing above. also learned the same way but yeah an big example covering basic concepts would help newbies. it can also help reduce issues & discussions about valtio usage as its so simple.

**Reply by dai-shi** [reply]
Would anyone help on this idea, example based documentation?

**Reply by deadcoder0904** [reply]
i would've but busy with many other projects at the moment. maybe someday i'll write an example-based blog but no promises. firmly believe this should be taught with react...much more simpler api than `useContext` lol :)

**Reply by dsacramone** [reply]
I think a great start, would be to create/document a "list" of useful/gotchas/topic.... THEN you tackle the individual items in the list with documentation/examples. So, even though people don't have the time to commit to working on the docs, perhaps they can update the list.. and maybe others can tackle. 

This way, we wouldn't need to "write out all the document" at once, but create & grow organically with an expanding documentation. This topic here, an example. This would be in the list:

1...
2....
3. ~~What is the difference between snap & state, and when you would use one or the other?~~
4....

Now item 3, is updated and added to the doc and noted as "complete".

**Reply by santiagopuentep** [reply]
I think that could work. I have my own place where I have gotchas, so I could post at least those.

**Comment by sankarkumar23** [comment]
I am still confused.. Here is what I am attempting to accomplish, but I haven't been successful in doing so
I've a reusable component and I used it multiple pages.. I've a businessId in valtio proxy and can be changed in a root component.. for ex: TopBar.tsx (dropdown with businesses).
Even if I do not change the businessId the useEffect inside the component is triggering and logging messages to the console whenever the component is rendered. My objective is to ensure that the useEffect is invoked exclusively when dropdown value is changed in TopBar.tsx.

Do I need to use subscribe for this?

In store.ts
```
const state = proxy<AppState>({
  currentUser: null, // { id: '00000000-0000-0000-0000-000000000000', },
  businessId: '', //
  isMobile: false,
});
```

In ResusableComponent.tsx
```
export default function ResuableComponent() {
  const snap = useSnapshot(state);
  useEffect(() => {
    // need to make a api call here and store the result in local state not in valtio proxy
    console.log("business id changed (state)");
  }, [state.businessId]);

  //If I understood it correctly we should not use snap.busineesId here.. snap should used only in render/return
  useEffect(() => {
    console.log("business id changed (snap)");
  }, [snap.businessId]);
}
```

**Reply by dai-shi** [reply]
My suggestion would be to destructure snap in render:
```js
  const { businessId } = snap;
  useEffect(() => {
    console.log("business id changed");
  }, [businessId]);
```

**Comment by sankarkumar23** [comment]
Even with destructure also the same result

```
const { businessId } = useSnapshot(state);
useEffect(() => {
    console.log("business id changed (snap)");
  }, [businessId]);
```

**Reply by dai-shi** [reply]
Does it behave unexpectedly?

**Comment by sankarkumar23** [comment]
Oh.. I think I figured it out.. I am navigating from one page to another page.. which is mounting and unmounting the reusable component.. so it makes sense console.log logging the message.. Any other techniques how to prevent this?

**Reply by dai-shi** [reply]
Maybe use `subscribe` instead of `useEffect`?

**Reply by sankarkumar23** [reply]
Yes,, subscribeKey does the trick. Thanks!
